File: typescript/src/step4-task-manager.ts
// Step 4: TaskManager クラスを作る
// 学ぶ概念：コンポジション、has-a 関係
//
// 実行方法:
//   npm run step4

// Task クラスは完成形を提供しています
class Task {
  public title: string;
  public dueDate: string;
  private _completed: boolean;

  constructor(title: string, dueDate: string) {
    this.title = title;
    this.dueDate = dueDate;
    this._completed = false;
  }

  get completed(): boolean {
    return this._completed;
  }

  set completed(value: boolean) {
    if (this._completed && !value) {
      throw new Error("完了済みのタスクを未完了に戻すことはできません");
    }
    this._completed = value;
  }

  complete(): void {
    this._completed = true;
  }

  display(): void {
    const status = this._completed ? "完了" : "未完了";
    console.log(`[${status}] ${this.title}（期日: ${this.dueDate}）`);
  }
}

// TODO: TaskManager クラスを実装する
//
// TaskManager は複数の Task をまとめて管理するクラスです
//
// プロパティ:
//   - tasks: Task の配列（private）
//
// メソッド:
//   - addTask(task: Task): void
//       tasks に Task を追加する
//
//   - removeTask(title: string): void
//       タイトルが一致する Task を tasks から削除する
//       ヒント: this.tasks = this.tasks.filter(...) で絞り込める
//
//   - findTask(title: string): Task | undefined
//       タイトルが一致する Task を返す。見つからなければ undefined を返す
//       ヒント: find() は filter() と違い「最初の1件だけ」を返す
//              見つからない場合は undefined になるため、戻り値の型が Task | undefined
//
//   - displayAll(): void
//       tasks の全タスクを display() で出力する
//
//   - get completedCount(): number  （getter）
//       完了済みタスクの件数を返す
//       ヒント: Step 3 で使った get キーワードと、filter().length の組み合わせ
//
//   - get pendingCount(): number  （getter）
//       未完了タスクの件数を返す

// --- 動作確認 ---
const manager = new TaskManager();

manager.addTask(new Task("買い物", "2024-12-31"));
manager.addTask(new Task("読書", "2024-11-30"));
manager.addTask(new Task("運動", "2024-10-15"));

console.log("--- 全タスク ---");
manager.displayAll();
// 期待:
// [未完了] 買い物（期日: 2024-12-31）
// [未完了] 読書（期日: 2024-11-30）
// [未完了] 運動（期日: 2024-10-15）

const found = manager.findTask("読書");
found?.complete();

manager.removeTask("運動");

console.log("\n--- 更新後 ---");
manager.displayAll();
// 期待:
// [未完了] 買い物（期日: 2024-12-31）
// [完了] 読書（期日: 2024-11-30）

console.log(`完了: ${manager.completedCount}件 / 未完了: ${manager.pendingCount}件`);
// 期待: 完了: 1件 / 未完了: 1件

export {};

File: python/src/step4_task_manager.py
# Step 4: TaskManager クラスを作る
# 学ぶ概念：コンポジション、has-a 関係
#
# 実行方法:
#   python src/step4_task_manager.py

from __future__ import annotations

# Task クラスは完成形を提供しています
class Task:
    def __init__(self, title: str, due_date: str) -> None:
        self.title = title
        self.due_date = due_date
        self._completed: bool = False

    @property
    def completed(self) -> bool:
        return self._completed

    @completed.setter
    def completed(self, value: bool) -> None:
        if self._completed and not value:
            raise ValueError("完了済みのタスクを未完了に戻すことはできません")
        self._completed = value

    def complete(self) -> None:
        self._completed = True

    def display(self) -> None:
        status = "完了" if self._completed else "未完了"
        print(f"[{status}] {self.title}（期日: {self.due_date}）")


# TODO: TaskManager クラスを実装する
#
# TaskManager は複数の Task をまとめて管理するクラスです
#
# プロパティ:
#   - __tasks: Task のリスト（private）、初期値は空リスト
#
# メソッド:
#   - add_task(task: Task) -> None
#       __tasks に Task を追加する
#
#   - remove_task(title: str) -> None
#       タイトルが一致する Task を __tasks から削除する
#       ヒント: リスト内包表記で新しいリストを作り直す
#              self.__tasks = [t for t in self.__tasks if t.title != title]
#
#   - find_task(title: str) -> Task | None
#       タイトルが一致する Task を返す。見つからなければ None を返す
#       ヒント: next((t for t in self.__tasks if t.title == title), None)
#              TypeScript の find() に相当する（見つからない場合は None = undefined と同等）
#
#   - display_all() -> None
#       __tasks の全タスクを display() で出力する
#
#   @property
#   - completed_count -> int（getter）
#       完了済みタスクの件数を返す
#
#   @property
#   - pending_count -> int（getter）
#       未完了タスクの件数を返す


# --- 動作確認 ---
manager = TaskManager()

manager.add_task(Task("買い物", "2024-12-31"))
manager.add_task(Task("読書", "2024-11-30"))
manager.add_task(Task("運動", "2024-10-15"))

print("--- 全タスク ---")
manager.display_all()
# 期待:
# [未完了] 買い物（期日: 2024-12-31）
# [未完了] 読書（期日: 2024-11-30）
# [未完了] 運動（期日: 2024-10-15）

found = manager.find_task("読書")
if found:
    found.complete()

manager.remove_task("運動")

print("\n--- 更新後 ---")
manager.display_all()
# 期待:
# [未完了] 買い物（期日: 2024-12-31）
# [完了] 読書（期日: 2024-11-30）

print(f"完了: {manager.completed_count}件 / 未完了: {manager.pending_count}件")
# 期待: 完了: 1件 / 未完了: 1件
